lorawan/src/subnet.rs

```diff
@@ -0,0 +1,386 @@
+pub fn is_local_devaddr(devaddr: u32, netid_list: Vec<u32>) -> bool {
+    let netid = parse_netid(devaddr);
+    is_local_netid(netid, netid_list)
+}
+
+pub fn devaddr_from_subnet(subnetaddr: u32, netid_list: Vec<u32>) -> u32 {
+    let netid = subnet_addr_to_netid(subnetaddr, netid_list.clone());
+    let (lower, _upper) = netid_addr_range(netid, netid_list);
+    devaddr(netid, subnetaddr - lower)
+}
+
+pub fn subnet_from_devaddr(devaddr: u32, netid_list: Vec<u32>) -> u32 {
+    let netid = parse_netid(devaddr);
+    let (lower, _upper) = netid_addr_range(netid, netid_list);
+    lower + nwk_addr(devaddr)
+}
+
+fn netid_class(netid: u32) -> u32 {
+    netid >> 21
+}
+
+fn addr_len(netclass: u32) -> u32 {
+    let result: u32 = match netclass {
+        0 => 25,
+        1 => 24,
+        2 => 20,
+        3 => 17,
+        4 => 15,
+        5 => 13,
+        6 => 10,
+        7 => 7,
+        _ => 0,
+    };
+    result
+}
+
+#[allow(dead_code)]
+fn addr_bit_len(devaddr: u32) -> u32 {
+    let netid = parse_netid(devaddr);
+    addr_len(netid_class(netid))
+}
+
+fn id_len(netclass: u32) -> u32 {
+    let result: u32 = match netclass {
+        0 => 6,
+        1 => 6,
+        2 => 9,
+        3 => 11,
+        4 => 12,
+        5 => 13,
+        6 => 15,
+        7 => 17,
+        _ => 0,
+    };
+    result
+}
+
+fn subnet_addr_to_netid(subnetaddr: u32, netid_list: Vec<u32>) -> u32 {
+    for item in netid_list.clone() {
+        if subnet_addr_within_range(subnetaddr, item, netid_list.clone()) {
+            return item;
+        }
+    }
+    0
+}
+
+fn subnet_addr_within_range(subnetaddr: u32, netid: u32, netid_list: Vec<u32>) -> bool {
+    let (lower, upper) = netid_addr_range(netid, netid_list);
+    (subnetaddr >= lower) && (subnetaddr < upper)
+}
+
+fn var_net_class(netclass: u32) -> u32 {
```

Comment:
would it be worth making a special enum type for netclass?
```rust
enum NetClass {
  _0,
  ...
  _7,
}
```

Reply:
Technically the netclass is a 3 bit value, i.e. u3.  I explored the enum idea, but netclass is used in different ways in the code, i.e. a number, a bitshift input value and bit pattern.  The code starts looking ugly when it is declared an enum.

I changed it to a u8.  The usage is safer now.  What do you think?

Reply:
> For example, all the functions that takes NetID List seem pretty domain specific and probably shouldn't be here (perhaps in the router module somehwere).

Good point and I agree with the domain specific premise.  Pragmatically though this same logic is shared between the light gateway, Miner, Router and PacketPurchaser repos.  Prior reviewers noted the duplicate code and consensus was to put the subnet translation logic into the shared LoRaWAN library.

Reply:
F# has an amazing type system, where for example, a custom type can be defined for two different unsigned ints and the compiler will enforce it.  The meter to feet conversion is the common example where you want to declare a distance in meters and distance in feet types and let the compiler strictly enforce it, so mistaken assignments do not occur.

Prof pointed out that something similar may exist for Rust
https://yoric.github.io/post/uom.rs/

Reply:
yeah.. I wouldn't try to overcomplicate things though